src/components/colorscale/calc.js

```diff
@@ -20,7 +20,7 @@ module.exports = function calc(trace, vals, containerStr, cLetter) {
 
     if(containerStr) {
         container = Lib.nestedProperty(trace, containerStr).get();
-        inputContainer = Lib.nestedProperty(trace._input, containerStr).get();
+        inputContainer = Lib.nestedProperty(trace, containerStr).get();
```

Comment:
> What we should do is: mark those attributes with a special flag in the attributes.js

I see what you mean, wondered why `marker.cauto` restyle magically worked and `line.cauto` didn't. Sang a mantra of relief when I found it :-) https://github.com/plotly/plotly.js/blob/master/src/plot_api/plot_api.js#L1559-L1594

> I don't remember one off the top of my head.

I switched the code as discussed. But it seems that if we persist the autocalculated `cmin`/`cmax` then it does *not* do what you want (provided I understand it well). Why? Only # 5 differs from my above list, the rest is copied.

1. Initially the plot has specific `cmin`/`cmax` values; `cauto` false
2. I do `Plotly.restyle(gd, 'marker.cauto', true)`
3. Result is, the color palette now expands to the entire range of the data as it should
4. I do `Plotly.restyle(gd, 'marker.cauto', false)`
5. Result is, the colors on the markers *do not get restored* to what they were because step 2 caused a destructive, irrecoverable mutation of the original user input

I reverted to my simpler code for now, haven't pushed the code snippet in the above comment but open to do whatever our clarification leads to.

> backward-incompatible change

When you mentioned this reliance I initially also thought that retaining compatibility goes without saying. But now there's this nagging feeling that maybe it's not a big risk to break compat for this specifically, and perhaps see if plot.ly/plot workspace needs fixing and handle reported issues in the (perhaps) unlikely case of user reliance. In general I'd favor retaining original user input and not sharing our own intermediate work calculations unless through a querying API (rather than directly querying objects which are implementation detail). As you have a lot more info on the consequences, of course its your call but wanted to share a base sentiment.